src/proxyConnect.ts

```diff
@@ -0,0 +1,132 @@
+import { ENCODING } from './xdebugConnection'
+
+export const DEFAULTIDEKEY = 'vsc'
+export interface ProxyMessages {
+    defaultError: string
+    deregisterInfo: string
+    deregisterSuccess: string
+    duplicateKey: string
+    nonexistentKey: string
+    registerInfo: string
+    registerSuccess: string
+    resolve: string
+    timeout: string
+}
+
+/** Informs proxy of incoming connection and who to pass data back to. */
+export class ProxyConnect extends EventEmitter {
+    /** Port editor is listening on (default: 9001 */
+    private _port: number
+    /** a CLI binary boolean option (default: 1) */
+    private _allowMultipleSessions: number
+    /** host domain or ip (default: 127.0.0.1) */
+    private _host: string
+    /** ide port proxy will connect back */
+    private _ideport: number
+    /** unique key that allows the proxy to match requests to your editor. (default: DEFAULTIDEKEY) */
+    private _key: string
+    /** proxy response data parser */
+    private _parser = new DOMParser()
+    /** tcp connection to communicate with proxy server */
+    private _socket: Socket
+    /** milliseconds to wait before giving up */
+    private _timeout: number
+    public msgs: ProxyMessages
+    private _isRegistered = false
+    private _resolve: Function
+
+    constructor(
+        host = '127.0.0.1',
+        port = 9001,
+        ideport = 9000,
+        allowMultipleSessions = true,
+        key = DEFAULTIDEKEY,
+        timeout = 3000,
+        socket?: Socket
+    ) {
+        super()
+        this._allowMultipleSessions = allowMultipleSessions ? 1 : 0
+        this._host = host
+        this._key = key
+        this._port = port
+        this._ideport = ideport
+        this._timeout = timeout
+        this._socket = !!socket ? socket : new Socket()
+        this.msgs = {
+            defaultError: 'Unknown Error',
+            deregisterInfo: `Deregistering ${this._key} with proxy @ ${this._host}:${this._port}`,
+            deregisterSuccess: 'Deregistration successful',
+            duplicateKey: 'IDE Key already exists',
+            nonexistentKey: 'No IDE key',
+            registerInfo: `Registering ${this._key} on port ${this._ideport} with proxy @ ${this._host}:${this._port}`,
+            registerSuccess: 'Registration successful',
+            resolve: `Failure to resolve ${this._host}`,
+            timeout: `Timeout connecting to ${this._host}:${this._port}`,
+        }
+        this._socket.on('error', (err: Error) => {
+            // Propagate error up
+            this._socket.end()
+            this.emit('error', err instanceof Error ? err : new Error(err))
+        })
+        this._socket.on('lookup', (err: Error | null, address: string, family: string | null, host: string) => {
+            if (err instanceof Error) {
+                this._socket.emit('error', `${err.message}${address || host || ''}`)
+            }
+        })
+        this._socket.on('data', data => this._responseStrategy(data))
```

Comment:
This assumes that response messages correspond 1:1 to `data` events, which is not true. It may be true often, especially when testing locally, but `data` events can be arbitrary chunks of data that may contain only part of a response. The `DBGPConnection` class handles this properly.

Reply:
As said, the IDE registration protocol is not the same as DBGP. The fact is this protocol lacks proper framing of the response (the length NULL xml NULL) so there is no way to know where the response ends. Also the response should always be way below one TCP segment... However, as all DBGP proxy implementations I could find work in a way that they close the TCP session immediately after sending the XML, I have changed the implementation so that it collects all 'data' `Buffer`s and process them on 'close' event.